pkg/agent/stats/collector.go

```diff
@@ -0,0 +1,208 @@
+	npStatsMap := map[types.UID]*metricsv1alpha1.TrafficStats{}
+	acnpStatsMap := map[types.UID]*metricsv1alpha1.TrafficStats{}
+	anpStatsMap := map[types.UID]*metricsv1alpha1.TrafficStats{}
+	for ofID, ruleStats := range ruleStatsMap {
+		policyRef := m.ofClient.GetPolicyFromConjunction(ofID)
+		// Same as above, this may be because the NetworkPolicy is removed right after the metrics are fetched.
+		if policyRef == nil {
+			klog.Infof("Cannot find NetworkPolicy that has ofID %v", ofID)
+			continue
+		}
+		klog.V(4).Infof("Converting ofID %v to policy %s", ofID, policyRef.ToString())
+
+		var statsMap map[types.UID]*metricsv1alpha1.TrafficStats
+		switch policyRef.Type {
+		case cpv1beta1.K8sNetworkPolicy:
+			statsMap = npStatsMap
+		case cpv1beta1.AntreaClusterNetworkPolicy:
+			statsMap = acnpStatsMap
+		case cpv1beta1.AntreaNetworkPolicy:
+			statsMap = anpStatsMap
+		}
+
+		policyStats, exists := statsMap[policyRef.UID]
+		if !exists {
+			policyStats = new(metricsv1alpha1.TrafficStats)
+			statsMap[policyRef.UID] = policyStats
+		}
+		policyStats.Bytes += int64(ruleStats.Bytes)
+		policyStats.Sessions += int64(ruleStats.Sessions)
+		policyStats.Packets += int64(ruleStats.Packets)
+	}
+	return &statsCollection{
+		networkPolicyStats:              npStatsMap,
+		antreaClusterNetworkPolicyStats: acnpStatsMap,
+		antreaNetworkPolicyStats:        anpStatsMap,
+	}
+}
+
+// report calculates the delta of the stats and pushes it to the antrea-controller summary API.
+func (m *Collector) report(curStatsCollection *statsCollection) error {
+	npStats := calculateDiff(curStatsCollection.networkPolicyStats, m.lastStatsCollection.networkPolicyStats)
+	acnpStats := calculateDiff(curStatsCollection.antreaClusterNetworkPolicyStats, m.lastStatsCollection.antreaClusterNetworkPolicyStats)
+	anpStats := calculateDiff(curStatsCollection.antreaNetworkPolicyStats, m.lastStatsCollection.antreaNetworkPolicyStats)
+	if len(npStats) == 0 && len(acnpStats) == 0 && len(anpStats) == 0 {
+		klog.V(4).Info("No stats to report, skip reporting")
+		return nil
+	}
+
+	summary := &cpv1alpha1.NodeStatsSummary{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: m.nodeName,
+		},
+		NetworkPolicies:              npStats,
+		AntreaClusterNetworkPolicies: acnpStats,
+		AntreaNetworkPolicies:        anpStats,
+	}
+	klog.V(6).Infof("Reporting NodeStatsSummary: %v", summary)
+
+	antreaClient, err := m.antreaClientProvider.GetAntreaClient()
+	if err != nil {
+		return err
+	}
+	_, err = antreaClient.ControlplaneV1alpha1().NodeStatsSummaries().Create(context.TODO(), summary, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func calculateDiff(curStatsMap, lastStatsMap map[types.UID]*metricsv1alpha1.TrafficStats) []cpv1alpha1.NetworkPolicyStats {
+	if len(curStatsMap) == 0 {
+		return nil
+	}
+	statsList := make([]cpv1alpha1.NetworkPolicyStats, 0, len(curStatsMap))
+	for uid, curStats := range curStatsMap {
+		var stats *metricsv1alpha1.TrafficStats
```

Comment:
After you asked this, I did a benchmark of calculating maps of 100 items and tried moving stats out of the loop. The performance became worse:
```
Before (stats is declared in the loop):
BenchmarkCalculateDiff-4           88396             11668 ns/op            9472 B/op          1 allocs/op

After (stats is declared out of the loop):
BenchmarkCalculateDiff-4           73246             16439 ns/op           12672 B/op        101 allocs/op
```
It might be related to where the pointer and struct variable is allocated, stack or heap. If the pointer and struct is both declared in the loop and it's finally copied to an allocated slice, it won't used outside of the loop so they are allocated on stack. If the pointer is declared out of the loop, go runtime may not know whether it will be used outside the loop so allocate the struct on heap.  See https://golang.org/doc/faq#stack_or_heap

Reply:
thanks this good read.. learn something new always